---
title: "Trajecctory inference of recruited lung monocytes & macrophages"
author: "Anson Wong"
date: '03-Dec-2023'
output: html_document
---

Load packages
```{r}
suppressPackageStartupMessages(library(monocle3))
suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(SeuratData))
suppressPackageStartupMessages(library(SeuratWrappers))
suppressPackageStartupMessages(library(patchwork))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(scCustomize))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(Polychrome))
suppressPackageStartupMessages(library(viridis))
suppressPackageStartupMessages(library(pals))
suppressPackageStartupMessages(library(nrmisc))
suppressPackageStartupMessages(library(paletteer))
suppressPackageStartupMessages(library(BiocStyle))
suppressPackageStartupMessages(library(here))
suppressPackageStartupMessages(library(glue))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(ggpubr))
suppressPackageStartupMessages(library(here))
suppressPackageStartupMessages(library(ggrepel))
suppressPackageStartupMessages(library(scales))
suppressPackageStartupMessages(library(ggsci))
suppressPackageStartupMessages(library(ggthemes))
suppressPackageStartupMessages(library(gplots))
suppressPackageStartupMessages(library(speckle))
suppressPackageStartupMessages(library(circlize))
suppressPackageStartupMessages(library(ComplexHeatmap))
suppressPackageStartupMessages(library(tidyHeatmap))
suppressPackageStartupMessages(library(ggplotify))
suppressPackageStartupMessages(library(jpeg))
suppressPackageStartupMessages(library(circlize))
suppressPackageStartupMessages(library(phateR))
suppressPackageStartupMessages(library(dittoSeq))
suppressPackageStartupMessages(library(circlize))
suppressPackageStartupMessages(library(ComplexHeatmap))
suppressPackageStartupMessages(library(tidyHeatmap))
suppressPackageStartupMessages(library(ggplotify))
suppressPackageStartupMessages(library(jpeg))
suppressPackageStartupMessages(library(circlize))
suppressPackageStartupMessages(library(slingshot))
suppressPackageStartupMessages(library(tradeSeq))
suppressPackageStartupMessages(library(tidymodels))
suppressPackageStartupMessages(library(reticulate))

```

Read object
```{r}
seu <- readRDS('data/SCEs/G000323.clustered.Zilionis.HLCA.mac.integrated.subclustered.annotated.SEU.rds')
seu$predicted.ann_level_3 <- factor(seu$predicted.ann_level_3)
seu$predicted.ann_level_4 <- factor(seu$predicted.ann_level_4)
seu$predicted.ann_finest_level <- factor(seu$predicted.ann_finest_level)
seu$Annotation <- factor(seu$Annotation)
seu$HTO_HTODemux <- factor(seu$HTO_HTODemux,
                           levels=paste0("Human_HTO_",c(3,6,7,8,10,12,13,14,15,16)))
# add condition1
condition1 <- setNames(c(rep("CF",8),
                         rep("Healthy",2)), levels(seu$HTO_HTODemux))
condition1 <- condition1[seu$HTO_HTODemux]
names(condition1) <- colnames(seu)
seu <- AddMetaData(seu, metadata = condition1, col.name = "condition1")
seu$condition1 <- as.factor(seu$condition1)

# add condition2
condition2 <- setNames(c(rep("ETI-treated",4), # Trikafta indiv as treated
                         rep("CF-untreated",4),
                         rep("Healthy",2)), levels(seu$HTO_HTODemux))
condition2 <- condition2[seu$HTO_HTODemux]
names(condition2) <- colnames(seu)
seu <- AddMetaData(seu, metadata = condition2, col.name = "condition2")
seu$condition2 <- factor(seu$condition2,
                         levels=c("ETI-treated","CF-untreated","Healthy"))

# add condition3
condition3 <- setNames(c(rep(c("ETI-treated_t1","ETI-treated_t2"),2),
                         rep(c("CF-untreated_t1","CF-untreated_t2"),2),
                         rep("Healthy",2)), levels(seu$HTO_HTODemux))
condition3 <- condition3[seu$HTO_HTODemux]
names(condition3) <- colnames(seu)
seu <- AddMetaData(seu, metadata = condition3, col.name = "condition3")
seu$condition3 <- as.factor(seu$condition3)

# add timepoint to metadata
timepoint <- setNames(c(rep(c("T1","T2"),4),
                         rep("N/A",2)), levels(seu$HTO_HTODemux))
timepoint <- timepoint[seu$HTO_HTODemux]
names(timepoint) <- colnames(seu)
seu <- AddMetaData(seu, metadata = timepoint, col.name = "Timepoint")
seu$Timepoint <- factor(seu$Timepoint,
                             levels=c("T1","T2","N/A"))

# reorder Annotation
seu$Annotation <- factor(seu$Annotation, levels=c("alvMΦ",
                                                  "recMo/MΦ",
                                                  "alvMΦ-proliferating",
                                                  "alvMΦ-CCL",
                                                  "alvMΦ-IFN",
                                                  "alvMΦ-lipid",
                                                  "alvMΦ-MT",
                                                  "alvMΦ-T"))

# add subject to metadata
Subject <- setNames(c("ETI1","ETI1",
                      "ETI2","ETI2",
                      "UT1","UT1",
                      "UT2","UT2",
                      "HC1","HC2"),levels(seu$HTO_HTODemux))

Subject <- Subject[seu$HTO_HTODemux]
names(Subject) <- colnames(seu)
seu <- AddMetaData(seu, metadata = Subject, col.name = "Subject")
seu$Subject <- factor(seu$Subject,
                      levels=c("ETI1","ETI2",
                               "UT1","UT2",
                               "HC1","HC2"))

# add Sample to metadata
Sample <- setNames(c("M1C188_T1","M1C188_T2",
                      "M1C160_T1","M1C160_T2",
                      "M1C170_T1","M1C170_T2",
                      "M1C176_T1","M1C176_T2",
                      "M1N092","M1N075"),levels(seu$HTO_HTODemux))

Sample <- Sample[seu$HTO_HTODemux]
names(Sample) <- colnames(seu)
seu <- AddMetaData(seu, metadata = Sample, col.name = "Sample")
seu$Sample <- factor(seu$Sample,
                          levels=c("M1C188_T1","M1C188_T2",
                      "M1C160_T1","M1C160_T2",
                      "M1C170_T1","M1C170_T2",
                      "M1C176_T1","M1C176_T2",
                      "M1N092","M1N075"))

# UMAP for Annotation - Fig. 1c
# make a flipped umap
flip.mtx <- seu[["umap"]][[]]
flip.mtx[,2] <- -flip.mtx[,2]
seu[['umap']] <- CreateDimReducObject(embeddings = flip.mtx, key = "UMAP_", global = T, assay = "RNA")

Annotation.colours <- setNames(c("#81a4cd", #AM
                             "#fb6f92", #recMo/MΦ
                             "#7d451b", #AM-proliferating
                             "#f77f00", #AM-CCL
                             "#b23a48", # AM-IFN
                             "#054a91", #AM-lipid
                             "#5a189a", #AM-MT
                             "#a3b18a" #AM-T
                             ), levels(seu$Annotation))

s <- seu
seu <- NULL

```

Dimensionality reduction with PHATE
```{r}
DefaultAssay(seu1) <- "RNA"
seu1 <- SCTransform(seu1, method = "glmGamPoi",
                   vars.to.regress = c("subsets_Mito_percent",
                                       "subsets_Ribo_percent"),
                   variable.features.n = 2000)

seurat_data <- as.data.frame(seu1@assays$RNA@data[VariableFeatures(seu1),])
phate_data <- t(seurat_data)

# when comparing CF_treated_t1 vs Healthy
# ref: https://rdrr.io/cran/phateR/f/inst/examples/bonemarrow_tutorial.Rmd
phate_output <- phate(phate_data,
                      ndim=2,
                      knn=50,
                      npca=5,
                      t=50
                      )

## stash the embeddings back into the seu1rat object as a dimension reduction object
embeddings <- phate_output$embedding

seu1[["PHATE"]] <- CreateDimReducObject(embeddings = embeddings, 
                                       key = "PHATE_", 
                                       assay = DefaultAssay(seu1))

# FindNeighbors
seu1 <- FindNeighbors(seu1, reduction = "PHATE", dims = 1:2)
seu1 <- FindClusters(seu1, resolution = 0.1, algorithm = 4, method="matrix")

Condition1.colors <- setNames(c("#a4133c","#5b8e7d"),
                              c("CF","Healthy"))
Condition2.colors <- setNames(c("#a4133c", "#f4d58d", "#5b8e7d"),
                              c("ETI-treated","CF-untreated","Healthy"))


seu1@reductions$PHATE@cell.embeddings[,c(1,2)] <- -seu1@reductions$PHATE@cell.embeddings[,c(1,2)]


# Slingshot
# Trajectory inference using Slingshot =========================================
# convert seu1 to sce
sce1 <- as.SingleCellExperiment(seu1)
## Remarks on counts
# note that pseu13dotime analysis is inferred from PHATE embeddings not the counts
# counts are only used for TradeSeq
# so you don't have to specify 'data' slot here
DimPlot(seu1, group.by = "seurat_clusters", reduction="PHATE")

# Run Slingshot
set.seed(1)
sce1 <- slingshot(sce1,
                 clusterLabels = seu1$seurat_clusters,
                 start.clus="8", #  ** define the starting cluster here **
                 reducedDim = "PHATE",
                 allow.breaks=FALSE)

# prepare sds object
sds1 <- as.SlingshotDataSet(sce1)
sds1

# F2d ==========================================================================
Pseudotime <- setNames(sce1$slingPseudotime_1, colnames(counts(sce1)))
pal <- c("#E3F2FD", "#B8C9E1", "#8CA0C5", "#6176AA", "#354D8E", "#0A2472")

seu1 <- AddMetaData(seu1, Pseudotime, col.name = "Pseudotime")

f3b1 <- FeaturePlot(seu1,
                    reduction = "PHATE",
                    features = "Pseudotime",
                    pt.size = 0.3,
                    order = TRUE) +
  xlim(c(min(data.frame(seu1@reductions$PHATE@cell.embeddings)$PHATE_1),
         max(data.frame(seu1@reductions$PHATE@cell.embeddings)$PHATE_1))) +
  ylim(c(min(data.frame(seu1@reductions$PHATE@cell.embeddings)$PHATE_2),
         max(data.frame(seu1@reductions$PHATE@cell.embeddings)$PHATE_2))) +
  scale_color_gradientn(colours = pal) +
  theme_minimal() +
  theme(panel.border = element_blank(), 
        panel.grid = element_blank(),
        text = element_text(family = "Ariel"),
        axis.text=element_blank(),
        axis.title = element_blank(),
        plot.title = element_blank(),
        legend.text = element_text(size = 10),
        legend.title = element_text(face="bold", size=12)) + NoLegend()

f3b2 <- DimPlot(seu1, 
        reduction = "PHATE",
        group.by = "condition2",
        pt.size = 0.3,
        cols = Condition2.colors,
        shuffle = TRUE) +
  theme_minimal() +
  theme(panel.border = element_blank(), 
        panel.grid = element_blank(),
        text = element_text(family = "Ariel"),
        axis.text=element_blank(),
        axis.title = element_blank(),
        plot.title = element_blank(),
        legend.position = "right",
        legend.text = element_text(size = 10),
        legend.title = element_text(face="bold", size=12)) + NoLegend()

f3b3 <- DimPlot(seu1, 
        reduction = "PHATE",
        group.by = "predicted.ann_level_3",
        pt.size = 0.3,
        #cols = Condition2.colors,
        shuffle = TRUE) +
  scale_color_manual(values = c("#ffdab9","#ff8fab")) +
  theme_minimal() +
  theme(panel.border = element_blank(), 
        panel.grid = element_blank(),
        text = element_text(family = "Ariel"),
        axis.text=element_blank(),
        axis.title = element_blank(),
        plot.title = element_blank(),
        legend.position = "bottom",
        legend.text = element_text(size = 10),
        legend.title = element_text(face="bold", size=12)) + NoLegend()

```

Visualize PHATE and select the central cluster
```{r}
# Visualize PHATE_cluster. Look for the central cluster as starting point
DimPlot(seu,
        reduction = "PHATE",
        group.by = "seurat_clusters")
```

Trajectory inference using Slingshot
```{r}
# convert SEU to SCE
sce <- as.SingleCellExperiment(seu)
## Remarks on counts
# note that pseudotime analysis is inferred from PHATE embeddings not the counts
# counts are only used for TradeSeq
# so you don't have to specify 'data' slot here

# Run Slingshot
set.seed(1)
sce <- slingshot(sce,
                 clusterLabels = seu$seurat_clusters,
                 start.clus="4", #  ** define the starting cluster here **
                 reducedDim = "PHATE",
                 allow.breaks=FALSE)

# prepare sds object
sds <- as.SlingshotDataSet(sce)
sds

# save objects
RM.list <- list(seu,sce,assoRes,endRes)

pdt.left <- setNames(sce$slingPseudotime_1, colnames(counts(sce)))
pdt.right <- setNames(sce$slingPseudotime_2, colnames(counts(sce)))

seu <- AddMetaData(seu, pdt.left, col.name = "pdt.left")
seu <- AddMetaData(seu, pdt.right, col.name = "pdt.right")

df <- data.frame("PHATE_1" = as.numeric(seu@reductions$PHATE@cell.embeddings[,1]),
                 "PHATE_2" = as.numeric(seu@reductions$PHATE@cell.embeddings[,2]),
                 "pdt.left" = as.numeric(seu@meta.data$pdt.left),
                 "pdt.right" = as.numeric(seu@meta.data$pdt.right))

f2d3 <- ggplot(df, aes(x=PHATE_1, y=PHATE_2)) +
  geom_point(aes(colour=pdt.left),size=0.3, alpha=1) +
  theme_classic() +
  scale_color_gradient2(high="#c9184a", low="#ffccd5", na.value = "#dedbd8") +
  scale_x_continuous(breaks = c(-0.01, 0.02, 0.05)) +
  scale_y_continuous(breaks = c(-0.001, 0.001, 0.003)) +
  theme(text = element_text(family = "Ariel"),
        axis.text.x = element_text(vjust = 1,
                                   hjust = 1,
                                   size=8,
                                   colour = "black"),
        axis.text.y = element_text(size=8,
                                   colour = "black"),
        plot.title = element_blank(),
        legend.position = "right",
        legend.text = element_text(size = 8),
        legend.title = element_text(face="bold", size=8)) +
  guides(fill = guide_legend(keyheight=0.1,
                             keywidth=0.1)) + NoLegend()

f2d4 <- ggplot(df, aes(x=PHATE_1, y=PHATE_2)) +
  geom_point(aes(colour=pdt.right),size=0.3, alpha=1) +
  theme_classic() +
  scale_color_gradient2(high="#0a2472", low="#e3f2fd", na.value = "#dedbd8") +
  scale_x_continuous(breaks = c(-0.01, 0.02, 0.05)) +
  scale_y_continuous(breaks = c(-0.001, 0.001, 0.003)) +
  theme(text = element_text(family = "Ariel"),
        axis.text.x = element_text(vjust = 1,
                                   hjust = 1,
                                   size=8,
                                   colour = "black"),
        axis.text.y = element_text(size=8,
                                   colour = "black"),
        plot.title = element_blank(),
        legend.position = "right",
        legend.text = element_text(size = 8),
        legend.title = element_text(face="bold", size=8)) +
  guides(fill = guide_legend(keyheight=0.1,
                             keywidth=0.1)) + NoLegend()

(f2d1+f2d2)/(f2d3+f2d4)

```

Differential expression using TradeSeq
```{r}
# prepare countMatrix ==========================================================
# Removing some genes to speed up the computations for this tutorial
# filt_counts <- counts
counts <- seu@assays$RNA@counts
filt_counts <- counts[rowSums(counts > 5) > ncol(counts)/100, ]

# # fit negative binomial model
# set.seed(1)
# icMat <- evaluateK(counts = filt_counts, sds = sds, k = 3:10, 
#                    nGenes = 200, verbose = T)

# GAM fitting
pseudotime <- slingPseudotime(sds, na = FALSE)
cellWeights <- slingCurveWeights(sds)
sce <- fitGAM(counts = filt_counts, 
              pseudotime = pseudotime, 
              cellWeights = cellWeights,
              nknots = 6, verbose = TRUE)

table(rowData(sce)$tradeSeq$converged)

# association of gene expression with pseudotime
assoRes <- associationTest(sce, global=FALSE, lineages=TRUE)
assoRes

# pdt.1
assoRes.pdt.1 <- data.frame(assoRes) %>% 
  dplyr::select(waldStat_1, df_1, pvalue_1) %>% 
  arrange(desc(waldStat_1))
# pdt.2
assoRes.pdt.2 <- data.frame(assoRes) %>% 
  dplyr::select(waldStat_2, df_2, pvalue_2) %>% 
  arrange(desc(waldStat_2))
# pdt.3
assoRes.pdt.3 <- data.frame(assoRes) %>% 
  dplyr::select(waldStat_3, df_3, pvalue_3) %>% 
  arrange(desc(waldStat_3))

head(assoRes.pdt.1,20)
head(assoRes.pdt.2,20)
head(assoRes.pdt.3,20)


# pdt.left
assoRes.pdt.left <- data.frame(assoRes) %>% 
  dplyr::select(waldStat_1, df_1, pvalue_1) %>% 
  arrange(desc(waldStat_1))
# pdt.right
assoRes.pdt.right <- data.frame(assoRes) %>% 
  dplyr::select(waldStat_2, df_2, pvalue_2) %>% 
  arrange(desc(waldStat_2))

assoRes.pdt.right

write.csv(assoRes.pdt.left,file = "assoRes.pdt.left.csv")
write.csv(assoRes.pdt.right,file = "assoRes.pdt.right.csv")


# between-lineage comparison
endRes <- diffEndTest(sce)
endRes %>% arrange(desc(waldStat)) %>% head(20)


# differential expression between starting point and end point of a lineage
startRes <- startVsEndTest(sce, lineages=TRUE)
head(startRes)

# pdt.1
startRes.pdt.1 <- data.frame(startRes) %>% 
  dplyr::select(waldStat_lineage1, pvalue_lineage1) %>% 
  arrange(desc(waldStat_lineage1))
# pdt.2
startRes.pdt.2 <- data.frame(startRes) %>% 
  dplyr::select(waldStat_lineage2, pvalue_lineage2) %>% 
  arrange(desc(waldStat_lineage2))
# pdt.3
startRes.pdt.3 <- data.frame(startRes) %>% 
  dplyr::select(waldStat_lineage3, pvalue_lineage3) %>% 
  arrange(desc(waldStat_lineage3))

head(startRes.pdt.1,20)
head(startRes.pdt.2,20)
head(startRes.pdt.3,20)

startRes %>% arrange(desc(waldStat))
class(startRes)
oStart <- order(startRes$waldStat, decreasing = TRUE)
oStart
sigGeneStart <- names(sce)[oStart[1]]
sigGeneStart
plotSmoothers(sce, counts, gene = "ITIH5")

(plotGeneCount(sds, counts, gene = "FABP4") + scale_color_gradientn(colors=inferno(256))) /
(plotGeneCount(sds, counts, gene = "APOC1") + scale_color_gradientn(colors=inferno(256))) /
(plotGeneCount(sds, counts, gene = "C1QB") + scale_color_gradientn(colors=inferno(256))) |
# data
((FeaturePlot(seu,
            reduction = "PHATE",
            features = c("FABP4")) +
  xlim(c(min(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_1),
         max(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_1))) +
  ylim(c(min(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_2),
         max(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_2))) +
  scale_color_gradientn(colours = inferno(256))) +
  (FeaturePlot(seu,
            reduction = "PHATE",
            features = c("APOC1")) +
  xlim(c(min(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_1),
         max(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_1))) +
  ylim(c(min(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_2),
         max(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_2))) +
  scale_color_gradientn(colours = inferno(256))) +
  (FeaturePlot(seu,
            reduction = "PHATE",
            features = c("C1QB")) +
  xlim(c(min(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_1),
         max(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_1))) +
  ylim(c(min(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_2),
         max(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_2))) +
  scale_color_gradientn(colours = inferno(256)))) |
# count
((FeaturePlot(seu, slot="counts",
            reduction = "PHATE",
            features = c("FABP4")) +
  xlim(c(min(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_1),
         max(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_1))) +
  ylim(c(min(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_2),
         max(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_2)))) +
  #scale_color_gradientn(colours = inferno(256))) +
  (FeaturePlot(seu, slot="counts",
            reduction = "PHATE",
            features = c("APOC1")) +
  xlim(c(min(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_1),
         max(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_1))) +
  ylim(c(min(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_2),
         max(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_2)))) +
  #scale_color_gradientn(colours = inferno(256))) +
  (FeaturePlot(seu, slot="counts",
            reduction = "PHATE",
            features = c("C1QB")) +
  xlim(c(min(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_1),
         max(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_1))) +
  ylim(c(min(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_2),
         max(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_2)))))
  #scale_color_gradientn(colours = inferno(256))))

pdt.matrix <- data.frame(pdt[,2])

cw.matrix[!is.na(cw.matrix)] <- 1
cw.matrix[is.na(cw.matrix)] <- 0
cw.matrix <- as.matrix(cw.matrix)

pdt.matrix[is.na(pdt.matrix)] <- 0
pdt.matrix <- as.matrix(pdt.matrix)

# GAM fitting
sce <- fitGAM(counts = as.matrix(filt_counts), 
              #sds = curves,
              pseudotime = pdt.matrix,
              cellWeights = cw.matrix)

# Define function to plot
plot_differential_expression <- function(feature_id) {
    # feature_id <- pseudotime_start_end_association %>% 
    #   dplyr::filter(pvalue < 0.05) %>% 
    #   dplyr::top_n(1, -waldStat) %>% 
    #   dplyr::pull(feature_id)
    cowplot::plot_grid(plotGeneCount(curves, filt_counts, 
                                     gene = feature_id[1], 
                                     clusters = clustering, models = sce) + 
                         ggplot2::theme(legend.position = "right") +
                         scale_color_gradientn(colours =inferno(256))) 
        #plotSmoothers(sce, as.matrix(counts), gene = feature_id[1]))
}

# DEGS that change with pseudotime
pseudotime_association <- associationTest(sce)
pseudotime_association$fdr <- p.adjust(pseudotime_association$pvalue, method = "fdr")
pseudotime_association <- pseudotime_association[order(pseudotime_association$pvalue), ]
pseudotime_association$feature_id <- rownames(pseudotime_association)

feature_id <- pseudotime_association %>% filter(pvalue < 0.05) %>% top_n(1, -waldStat) %>% pull(feature_id)
feature_id
plot_differential_expression(feature_id)

# DEGs that between two pseudotime points
pseudotime_start_end_association <- startVsEndTest(sce, 
                                                   pseudotimeValues = c(0, 1),
                                                   lineages=TRUE)
pseudotime_start_end_association$fdr <- p.adjust(pseudotime_start_end_association$pvalue, method = "fdr")
pseudotime_start_end_association <- pseudotime_start_end_association[order(pseudotime_start_end_association$pvalue), ]
pseudotime_start_end_association$feature_id <- rownames(pseudotime_start_end_association)

feature_id <- pseudotime_start_end_association %>% 
  filter(pvalue < 0.05) %>% 
  top_n(10, waldStat) %>% 
  pull(feature_id)

feature_id
feature_id <- c("CTSL","LYZ","LGALS1","C1QA")
plot_differential_expression(feature_id)



```

Plot PHATE
```{r}
Condition.colors <- setNames(c("#a4133c","#5b8e7d"),
                              levels(factor(seu$condition1)))

Subject.colors <- setNames(c(kelly(22)[c(11,16)],
                           kelly(22)[c(3,15)],
                           kelly(22)[c(19,10)]),
                           levels(factor(seu$Subject)))

Sample.colors <- setNames(kelly(22)[3:12],
                          levels(factor(seu$Sample)))

# plot using seurat's default tools
phate.df <- data.frame(seu@reductions$PHATE@cell.embeddings)
metdata.df <- seu@meta.data %>% dplyr::select(Subject, Sample, condition1,condition2,condition3)

df <- cbind(phate.df, metdata.df)
df$condition1 <- factor(df$condition1,
                        levels=c("CF","Healthy"))

# colored by gene expression
ggplot(df) +
  #geom_jitter(aes(PHATE_1, PHATE_2, color=df$MSR1)) +
  geom_point(aes(PHATE_1, PHATE_2, color=condition1),size=1,) +
  scale_color_manual(values = Condition.colors) +
  #scale_color_gradientn(colours = inferno(16)) +
  theme_classic() +
  theme()


# when comparing CF_treated_t1 vs Healthy
phate_output <- phate(phate_data,
                      ndim=2,
                      knn=50,
                      npca=5,
                      t=50
                      )

## stash the embeddings back into the seurat object as a dimension reduction object
seu[["PHATE"]] <- CreateDimReducObject(embeddings = phate_output$embedding, 
                                       key = "PHATE_", 
                                       assay = DefaultAssay(seu))
# plot using seurat's default tools
DimPlot(seu , 
        reduction = "PHATE",
        group.by = "condition3",
        #cols=Condition.colors,
        shuffle = TRUE) + ggtitle(label = "PHATE")

plot_features <- function(x) {
  FeaturePlot(seu,
            reduction = "PHATE",
            features = x) +
  xlim(c(min(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_1),
         max(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_1))) +
  ylim(c(min(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_2),
         max(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_2))) +
  scale_color_gradientn(colours = inferno(256))
}


plot_features("LYZ") + plot_features("CTSL") + plot_features("C1QA")


FeaturePlot(seu,
            reduction = "PHATE",
            features = c("LYZ","CTSL","C1QA")) +
  xlim(c(min(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_1),
         max(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_1))) +
  ylim(c(min(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_2),
         max(data.frame(seu@reductions$PHATE@cell.embeddings)$PHATE_2))) +
  scale_color_gradientn(colours = inferno(256))

```

Write plot
```{r}
pdf("pseudotime.DE.230707.pdf",height = 8,width=6)
par(mfrow = c(3, 4))
for (i in seq_along(top_genes)) {
  # colored by genes
  colors <- inferno(64)[cut(dat_use[,top_genes[i]], breaks = 64)]
  # colored by condition
  #colors <- Condition.colors[dat_use[,top_genes[i]]]
  plot(data.frame(reducedDim(sds)) %>% dplyr::arrange(top_genes[i]), col = colors, 
       pch = 18, cex = 1.5, main = top_gene_name[i])
  #lines(sds, lwd = 2, cex=0, col = 'black', type = 'lineages')
}
dev.off()

```

prepare object for heatmap
```{r}
# scale data
seu <- ScaleData(seu)

seu$HTO_HTODemux <- factor(seu$HTO_HTODemux,
                           levels=paste0("Human_HTO_",c(3,6,7,8,10,12,13,14,15,16)))

# add subject to metadata
Subject <- setNames(c("ETI1","ETI1",
                      "ETI2","ETI2",
                      "UT1","UT1",
                      "UT2","UT2",
                      "HC1","HC2"),levels(seu$HTO_HTODemux))

Subject <- Subject[seu$HTO_HTODemux]
names(Subject) <- colnames(seu)
seu <- AddMetaData(seu, metadata = Subject, col.name = "Subject")
seu$Subject <- factor(seu$Subject,
                          levels=c("ETI1","ETI2",
                                   "UT1","UT2",
                                   "HC1","HC2"))

# add Sample to metadata
Sample <- setNames(c("M1C188_T1","M1C188_T2",
                      "M1C160_T1","M1C160_T2",
                      "M1C170_T1","M1C170_T2",
                      "M1C176_T1","M1C176_T2",
                      "M1N092","M1N075"),levels(seu$HTO_HTODemux))

Sample <- Sample[seu$HTO_HTODemux]
names(Sample) <- colnames(seu)
seu <- AddMetaData(seu, metadata = Sample, col.name = "Sample")
seu$Sample <- factor(seu$Sample,
                          levels=c("M1C188_T1","M1C188_T2",
                      "M1C160_T1","M1C160_T2",
                      "M1C170_T1","M1C170_T2",
                      "M1C176_T1","M1C176_T2",
                      "M1N092","M1N075"))

# add condition1 to metadata
condition1 <- setNames(c(rep("CF",8), # Trikafta indiv as treated
                         rep("Healthy",2)), levels(seu$HTO_HTODemux))
condition1 <- condition1[seu$HTO_HTODemux]
names(condition1) <- colnames(seu)
seu <- AddMetaData(seu, metadata = condition1, col.name = "condition1")
seu$condition1 <- factor(seu$condition1)

# add Timepoint to metadata
timepoint <- setNames(c(rep(c("T1","T2"),4),
                         rep("N/A",2)), levels(seu$HTO_HTODemux))
timepoint <- timepoint[seu$HTO_HTODemux]
names(timepoint) <- colnames(seu)
seu <- AddMetaData(seu, metadata = timepoint, col.name = "Timepoint")
seu$Timepoint <- factor(seu$Timepoint,
                             levels=c("T1","T2","N/A"))

# Condition.colors <- setNames(c("#a4133c","#5b8e7d"),
#                               levels(seu$condition3))

Subject.colors <- setNames(c(kelly(22)[c(11,16)],
                           kelly(22)[c(3,15)],
                           kelly(22)[c(19,10)]),
                           levels(seu$Subject))

Timepoint.colors <- setNames(kelly(22)[c(6,12,9)],
                             levels(seu$Timepoint))

Sample.colors <- setNames(kelly(22)[3:12],
                          levels(seu$Sample))

```

Plot heatmap
```{r}
seu<- AddMetaData(seu, metadata = pdt, col.name = "pseudotime")

dittoHeatmap(seu,
             genes=top_genes,
             # genes=c("FABP5","GPNMB","CD9","CD81","APOC1",
             #         "LGMN","APOE","MSR1","MRC1","TREM2",
             #         "C1QA","C1QB","C1QC"),
             #order.by = c("MΦ.subtype","Condition","Sample"),
             order.by = ,
             annot.by = rev(c("pseudotime","condition3")),
             heatmap.colors.max.scaled = inferno(256),
             width=unit(0.0001/ncol(seu), "npc"),
             complex = TRUE,
             scaled.to.max = TRUE,
             cluster_cols = FALSE,
             cluster_rows = FALSE,
             # annotation_colors = list("condition3" = Condition.colors
             #                          ),
             #gaps_col=c(10,20,30,40,50,60,70),
             legend_breaks = c(0,0.2,0.4,0.6,0.8,1),
             name = "Scale",
             fontsize_row=8,
             fontface_row = "italic",
             # annotation_legend_param = list(labels_gp = gpar(fontface="italic",
             #                                fontsize=8)),
             heatmap_legend_param = list(labels_gp = gpar(fontsize = 8),
                                         legend_direction = "vertical"
                                         ),
             #cellheight=12,
             #cellwidth=10
             )


dittoHeatmap(seu.CF_treated_t2,
             genes=deg.top400[1:200],
             # genes=c("FABP5","GPNMB","CD9","CD81","APOC1",
             #         "LGMN","APOE","MSR1","MRC1","TREM2",
             #         "C1QA","C1QB","C1QC"),
             #order.by = c("MΦ.subtype","Condition","Sample"),
             order.by = c("CST3"),
             annot.by = rev(c("pseudotime","condition3")),
             heatmap.colors.max.scaled = inferno(256),
             width=unit(0.0001/ncol(seu), "npc"),
             complex = TRUE,
             scaled.to.max = TRUE,
             cluster_cols = FALSE,
             cluster_rows = FALSE,
             # annotation_colors = list("condition3" = Condition.colors
             #                          ),
             #gaps_col=c(10,20,30,40,50,60,70),
             legend_breaks = c(0,0.2,0.4,0.6,0.8,1),
             name = "Scale",
             fontsize_row=8,
             fontface_row = "italic",
             # annotation_legend_param = list(labels_gp = gpar(fontface="italic",
             #                                fontsize=8)),
             heatmap_legend_param = list(labels_gp = gpar(fontsize = 8),
                                         legend_direction = "vertical"
                                         ),
             #cellheight=12,
             #cellwidth=10
             )

```









